File: Leetcode/0173 Binary Search Tree Iterator.ts
class BSTIterator {
    private stack: TreeNode[] = [];
    
    constructor(root: TreeNode | null) {
        this.populateStack(root);
    }

    next(): number {
        const top = this.stack.pop() as TreeNode;
        this.populateStack(top.right);
        return top.val;
    }

    hasNext(): boolean {
        return this.stack.length !== 0;
    }

    private populateStack(root: TreeNode | null) {
        while (root !== null) {
            this.stack.push(root);
            root = root.left;
        }
    }
}

File: Leetcode/0173 Binary Search Tree Iterator.py
class BSTIterator:
    def __init__(self, root):
        self.stack = []
        self.__populate_stack(root)

    def next(self):
        top = self.stack.pop()
        self.__populate_stack(top.right)
        return top.val

    def hasNext(self):
        return len(self.stack) != 0

    def __populate_stack(self, root):
        while root:
            self.stack.append(root)
            root = root.left
